File: ruby/Roster.rb
require './Student.rb'
require "./Degree.rb"
class Roster
    @studentRoster
    @totalStudents

    def initialize()
           @studentRoster = Array[]
           @totalStudents = 0 
    end

    def studentRoster
      @studentRoster
    end

    def parseString(data)
          array = data.split(",")
          if(array[8] == "SECURITY")
            degree = Degree::SECURITY   
          elsif(array[8] == "NETWORK")
            degree = Degree::NETWORK  
          else
            degree = Degree::SOFTWARE
          end

          addStudent(array[0], array[1], array[2], array[3], array[4], [array[5], array[6], array[7]], degree)
    end

    def addStudent(id,  firstName,  lastName,  emailAddress,  age, coursedays,  degreeprogram)
            student = Student.new(id, firstName, lastName, emailAddress, age, degreeprogram, coursedays)
            @studentRoster.push(student)
            @totalStudents += 1
       
    end

    def removeStudent(id)
      located = false
      @studentRoster.each do |student|
        if(student.id() == id)
          @studentRoster.delete(student)
          @totalStudents -= 1
          located = true
        end  
      end
      if (located == false)
        puts "The student with the ID: " + id + " was not found." 
      end
    end
    
    def printAllStudents()
            @studentRoster.each do |student|
              student.Print()
            end
    end

    def printAverageDaysInCourse(id)
      totaldays = 0
      
      @studentRoster.each do |student|
        if(student.id() == id)
          student.CourseDays().each do |day|
            totaldays += Integer(day)
          end
        end  
      end
      averageDays = totaldays / 3
      puts "Student ID: " + id + ", Average days in course is: " + String(averageDays)
    end

    def printInvalidEmails()
      @studentRoster.each do |student|
       email = student.Email()
       if(email.include? " " or not email.include? "."or not email.include? "@")
                 puts email + " - Is invalid."
       end
      end
    end

    def printByDegreeProgram(degree)
      @studentRoster.each do |student|
        if student.Student_Degree() == degree 
          student.Print()
        end
      end
    end
end

File: php/Roster.php
<?php
require(__DIR__ ."/Student.php");

class Roster {
    public array $studentRoster = [];

    function parseString(string $data) {
        $array = explode(",", $data);
        $degree = null;

        if($array[8] == "SECURITY"){
            $degree = Degree::SECURITY;
        }elseif($array[8] == "NETWORK"){
            $degree = Degree::NETWORK;
        }else {
            $degree = Degree::SOFTWARE;
        }
       
        $this->addStudent($array[0], $array[1], $array[2], $array[3], $array[4], [$array[5], $array[6], $array[7]], $degree);
    }

    function addStudent(string $id, string $firstName, string $lastName, string $emailAddress, int  $age, array $coursedays, Degree $degreeprogram) {
    $student = new Student($id, $firstName, $lastName,  $emailAddress,   $age,  $degreeprogram, $coursedays);
    array_push($this->studentRoster, $student);
    }
    function removeStudent(string $id) {
        $located = false;
        for ($x = 0; $x < sizeof($this->studentRoster); $x++) {
            if ($this->studentRoster[$x]->getID() == $id) {
                array_splice($this->studentRoster, $x, 1);
                $located = true;
            }
          }
        
        if($located == false) {
            print("The student with the ID: " . $id . " was not found.");
        }
    }
    function printAllStudents() {
        foreach ($this->studentRoster as $student){
            $student->Print();
        }
    }
    function printAverageDaysInCourse(string $id) {
        $totalDays = 0;

        foreach ($this->studentRoster as $student){
           if($student->getID() == $id){
            foreach ($student->getCourseDays() as $day) {
                $totalDays += $day;
            }
           }   
          }
          $averageDays = $totalDays / 3;
          print("Student ID: " . $id . ", Average days in course is: " . $averageDays."\n");
    }
    function printInvalidEmails() {
        foreach ($this->studentRoster as $student){
          $email = $student->getEmail();
          if(str_contains($email, " ") || !str_contains($email, ".")|| !str_contains($email, "@")) {
            print($email . " - is invalid.\n");
          }
            
        }
    }
    function printByDegreeProgram(Degree $degree) {
        foreach ($this->studentRoster as $student){
            if($student->getDegree()  == $degree) {
                $student->Print();
            }
        }
    }
}